src/Microsoft.Health.Dicom.SqlServer/Features/Schema/Sql/Sprocs/EndUpdateInstanceV50.sql

```diff
@@ -0,0 +1,222 @@
+                OriginalWatermark BIGINT,
+                InstanceKey BIGINT)
+
+        DELETE FROM #UpdatedInstances
+
+        UPDATE dbo.Instance
+        SET LastStatusUpdatedDate = @currentDate,
+            OriginalWatermark = ISNULL(OriginalWatermark, Watermark),
+            Watermark = NewWatermark,
+            NewWatermark = NULL
+        OUTPUT deleted.PartitionKey, @studyInstanceUid, deleted.SeriesInstanceUid, deleted.SopInstanceUid, inserted.Watermark, inserted.OriginalWatermark, deleted.InstanceKey INTO #UpdatedInstances
+        WHERE PartitionKey = @partitionKey
+          AND StudyInstanceUid = @studyInstanceUid
+          AND Status = 1
+          AND NewWatermark IS NOT NULL
+
+        -- Create index on temp table so we can join on watermark when we need to update change feed with file path.
+        IF NOT EXISTS (SELECT *
+                       FROM   tempdb.sys.indexes
+                       WHERE  name = 'IXC_UpdatedInstances')
+            CREATE UNIQUE INDEX IXC_UpdatedInstances ON #UpdatedInstances (Watermark)
+
+        -- Create index on temp table so we can join on instance key and watermark when we need to insert file
+        -- properties when instance key alone is does not specify a unique row on the table
+        IF NOT EXISTS (SELECT *
+                       FROM   tempdb.sys.indexes
+                       WHERE  name = 'IXC_UpdatedInstanceKeyWatermark')
+        CREATE UNIQUE CLUSTERED INDEX IXC_UpdatedInstanceKeyWatermark ON #UpdatedInstances (InstanceKey, OriginalWatermark)
+
+        -- Only updating patient information in a study
+        UPDATE dbo.Study
+        SET PatientId = ISNULL(@patientId, PatientId),
+            PatientName = ISNULL(@patientName, PatientName),
+            PatientBirthDate = ISNULL(@patientBirthDate, PatientBirthDate),
+            @studyKey = StudyKey
+        WHERE PartitionKey = @partitionKey
+            AND StudyInstanceUid = @studyInstanceUid
+
+        -- The study does not exist. May be deleted
+        IF @@ROWCOUNT = 0
+            THROW 50404, 'Study does not exist', 1
+
+        -- Delete from file properties any rows with "stale" watermarks if we will be inserting new ones
+        IF EXISTS (SELECT 1 FROM @insertFileProperties)
+        DELETE FP
+        FROM dbo.FileProperty as FP
+        INNER JOIN #UpdatedInstances U
+        ON U.InstanceKey = FP.InstanceKey
+        WHERE U.OriginalWatermark != FP.Watermark
+
+        -- Insert new file properties from added blobs, @insertFileProperties will be empty when external store not
+        -- enabled
+        INSERT INTO dbo.FileProperty
+        (InstanceKey, Watermark, FilePath, ETag)
+        SELECT U.InstanceKey, I.Watermark, I.FilePath, I.ETag
+        FROM @insertFileProperties I
+        INNER JOIN #UpdatedInstances U
+        ON U.Watermark = I.Watermark
+
+        -- Update extended query tags value if any
+        -- String Key tags
+        IF EXISTS (SELECT 1 FROM @stringExtendedQueryTags)
```

Comment:
Should we just re-use `IIndexInstanceCoreV9` here instead?

Reply:
I thought about that, but it expects watermark which is an instance level thing, here I go and update for all study Key. Else I have call that sproc for all instances. Also that sproc merges when the TagStatus is either 0 or 1. But we dont want to update while it is creating, we want to do only when it is ready.

Reply:
using that procedure now. Make sense